notebooks/data_table_transformer/main.py

```diff
@@ -0,0 +1,125 @@
+
+with herzog.Cell("markdown"):
+    """
+    Carry out the long to wide data table transformation, which involves several steps:
+      - Convert the Terra data table into a Pandas dataframe with `table_to_dataframe`
+      - Filter out rows that cannot be transformed into long format. In this example we are transforming the data table
+        into a wide format using sample id. Multisample files, such as jointly called VCFs, will not fit into the
+        transformed dataframe. There are also extra complexities of the AnVIL data model that need to be addressed,
+        as explained below.
+      - Convert the wide-formatted dataframe into a Terra data table.
+    """
+
+import json
+import pandas
+from uuid import uuid4
+from random import choice, randint
+from terra_notebook_utils import table
+
+NUMBER_OF_WIDE_ROWS = 20
+
+long_table = "test-long"
+wide_table = "test-wide"
+
+table.delete(long_table)
+table.delete(wide_table)
+
+lines = list()
+for _ in range(NUMBER_OF_WIDE_ROWS):
+    sample_id = f"{uuid4()}"
+    sample_obj = dict(foo=f"{uuid4()}", entityName=sample_id)
+    for fmt in ["cram", "crai", "vcf"]:
+        lines.append({'pfb:sample': sample_obj,
+                      'pfb:data_format': fmt,
+                      'pfb:object_id': f"{uuid4()}",
+                      'pfb:file_size': randint(1024, 1024 ** 3),
+                      'pfb:file_name': f"{uuid4()}"})
+dataframe_to_table("test-long", pandas.DataFrame(lines))
+
+with herzog.Cell("python"):
+    index_column = "pfb:sample"
+    header_column = "pfb:data_format"
+    value_columns = ["pfb:object_id", "pfb:file_size", "pfb:file_name"]
+
+    # Uncomment the following lines with your table names
+    # long_table = ""  # Include long formatted table to transform
+    # wide_table = ""  # Include the wide formatted table to create
+
+    # Convert the long table into a Pandas dataframe
+    df = table_to_dataframe(long_table)
+    # Filter out rows without sample ids (such as multisample files)
+    df = df[df[index_column].notna()]
+    # "pfb:sample" is a complex field in AnVIL. Convert it to a string container the sample id
+    df[index_column] = df[index_column].apply(lambda item: item['entityName'])
+    # Use Pandas "pivot" method to convert  from long to wide
+    df_wide = df.pivot(index=index_column, columns=header_column, values=value_columns)
+    # Convert the wide dataframe into a Terra data table
+    dataframe_to_table(wide_table, df_wide)
+
+assert NUMBER_OF_WIDE_ROWS == len(table_to_dataframe(wide_table))
+
+with herzog.Cell("markdown"):
+    """
+    Alternatively to the above cell, [terra-pandas](https://github.com/xbrianh/terra-pandas) provides a convenience
+    function to carry out the long to wide data table transformation with a single command.
+    """
+
+with herzog.Cell("python"):
+    # long_to_wide("sequencing",
+    #              "sequencing-wide-two",
+    #              index_column="pfb:sample",
```

Comment:
Can you add "pfb:submitter_id" to the new data table and remove the weird file endings (.final.cram.crai, etc.)? Or bring that information from the sample data table? A non-UUID sample ID is a common input for workflows.

Reply:
Will add `fb:submitter_id`.

The weird file names are from Gen3, not this notebook, but I can remove `pfb:file_name` from the transformed table.

Reply:
Yeah, the not weird sample ID is in the `Sample` data table under `pfb:specimen_id`. The other admin nodes may also hold information that folks want to keep with the sequencing data for provenance (for example, if they copy only the sequencing-wide tsv to another workspace). In future iterations, depending on user feedback, we could consider another notebook example merging the `sequencing-wide` data table with the sample and project data tables (a la BDCat). I think we can wait on that though.

I think removing `pfb:file_name` is fine as you only need one representation of the sample ID provided by `pfb:submitter_id`.